File: Arrays/maximum_product_subarray.cpp
#include<iostream>
#include<vector>
#include<climits>
using namespace std;

int subarrays_max_product_optimal(vector<int> arr) {
    int prefix = 1, suffix = 1, maxi = INT_MIN;
    int n = arr.size();

    for (int i = 0; i < n; i++) {
        prefix = (prefix == 0) ? 1 : prefix;
        suffix = (suffix == 0) ? 1 : suffix;

        prefix *= arr[i];
        suffix *= arr[n - i - 1];

        maxi = max(maxi, max(prefix, suffix));
    }

    return maxi;
}


int subarrays_max_product_brute(vector<int> arr) {
    int n = arr.size();
    int maxi = INT_MIN;

    for (int i = 0; i < n; i++) {
        int prd = 1;
        for (int j = i; j < n; j++) {
            prd *= arr[j];
            maxi = max(maxi, prd);  
        }
    }

    return maxi;
}

int main() {
    vector<int> arr = {2, 3, -2, 4};  // Output should be 6
    cout << "Brute: " << subarrays_max_product_brute(arr) << endl;
    cout << "Optimal: " << subarrays_max_product_optimal(arr) << endl;
    return 0;
}

File: java/ArrayQ/MaximumProductSubarray.java
import java.util.*;

public class MaximumProductSubarray {

    /*
    ====================================================
    OPTIMAL APPROACH (Prefix & Suffix Product)
    ====================================================

    Intuition:
    ----------
    - Negative numbers can flip the sign of the product
    - Zero breaks subarrays
    - So we compute:
        1) Prefix product from left to right
        2) Suffix product from right to left
    - Reset product to 1 whenever it becomes 0
    - The maximum of prefix and suffix at each step
      gives the maximum product subarray

    Time Complexity:  O(n)
    Space Complexity: O(1)
    */
    public static int subarraysMaxProductOptimal(int[] arr) {
        int prefix = 1;
        int suffix = 1;
        int maxi = Integer.MIN_VALUE;
        int n = arr.length;

        for (int i = 0; i < n; i++) {
            if (prefix == 0) prefix = 1;
            if (suffix == 0) suffix = 1;

            prefix *= arr[i];
            suffix *= arr[n - i - 1];

            maxi = Math.max(maxi, Math.max(prefix, suffix));
        }
        return maxi;
    }

    /*
    ====================================================
    BRUTE FORCE APPROACH
    ====================================================

    Intuition:
    ----------
    - Try every possible subarray
    - Compute product and track maximum

    Time Complexity:  O(n^2)
    Space Complexity: O(1)
    */
    public static int subarraysMaxProductBrute(int[] arr) {
        int n = arr.length;
        int maxi = Integer.MIN_VALUE;

        for (int i = 0; i < n; i++) {
            int product = 1;
            for (int j = i; j < n; j++) {
                product *= arr[j];
                maxi = Math.max(maxi, product);
            }
        }
        return maxi;
    }

    /*
    ====================================================
    MAIN METHOD
    ====================================================
    */
    public static void main(String[] args) {
        int[] arr = {2, 3, -2, 4}; // Expected output: 6

        System.out.println("Brute: " + subarraysMaxProductBrute(arr));
        System.out.println("Optimal: " + subarraysMaxProductOptimal(arr));
    }
}
